README.md
# Thingboard API
Basic auth and get functions for Thingsboard REST API && Thingsboard's Postgres DB.
Promise like implementation



<!-- TOC -->

- [Installation](#installation)
- [Usage](#usage)
- [List REST API Functions](#listrestapifunctions)
    - [GET](#get)
        - [createConnection](#createconnection)
        - [get.objectID](#getobjectid)
        - [get.objectIDandKeys](#getobjectidandkeys)
        - [get.allObjectsIDbyType](#getallobjectsidbytype)
        - [get.allObjectsIDandKeysByType](#getallobjectsidandkeysbytype)
    - [PUSH](#push)
        - [push.createRelation](#pushcreaterelation)
        - [push.pushAttributes](#pushpushattributes)
        - [push.createEntity](#pushcreateentity)
    - [DELETE](#delete)
        - [delete.deleteEntity](#deletedeleteentity)
        - [delete.deleteChilds](#deletedeletechilds)
        - [delete.deleteEntitiesByType](#deletedeleteentitiesbytype)
- [List Postgres Functions](#listpostgresfunctions)
    - [postgres.get.allObjectsIDbyType](#postgresgetallobjectsidbytype)
    - [postgres.get.allObjectsIDandKeysByType](#postgresgetallobjectsidandkeysbytype)


<!-- /TOC -->

## Installation
```bash
npm i thingsboard_api
```

## Usage
```js
var TB = require('thingsboard_api');
var options = {
    TB_HOST:            'ip_to_TB',         //default: localhost
    TB_PORT:            'port_to_TB',       //default: 8080
    TB_USERNAME:        'loginTB',          //default: tenant@thingboard.org
    TB_PASSWORD:        'passTB',           //default: tenant 
    POSTGRES_HOST:      'ip_to_Postgres',   //no defaults!
    POSTGRES_PORT:      'port_to_Postgres', //no defaults!
    POSTGRES_USERNAME:  'loginPostgres',    //no defaults!
    POSTGRES_PASSWORD:  'loginPostgres',    //no defaults!

}

async function main(){
    await TB.createConnection(options) //
    console.log( await TB.get.objectID('entity_name','asset'))
    console.log( await TB.get.objectIDandKeys('entity_name2','device','key1,key2'))
}

main()

```


# List REST API Functions

## GET

### createConnection() 
### Promise connection with TB creating. Starting crone for token update every 15 minutes. NEED IN run once at start!

List of options:
``` js
options = {
    TB_HOST:            'ip_to_TB',         //default: localhost
    TB_PORT:            'port_to_TB',       //default: 8080
    TB_USERNAME:        'loginTB',          //default: tenant@thingboard.org
    TB_PASSWORD:        'passTB',           //default: tenant 
    POSTGRES_HOST:      'ip_to_Postgres',   //no defaults!
    POSTGRES_PORT:      'port_to_Postgres', //no defaults!
    POSTGRES_USERNAME:  'loginPostgres',    //no defaults!
    POSTGRES_PASSWORD:  'loginPostgres',    //no defaults!

}
```

Usage
```js
var TB = require('thingsboard_api');
await TB.createConnection(options)
```



### get.objectID()
### Promise get ID of an object by its name and type

List of options:
```json
(name, entity_view) //string
```

Usage
```js
var TB = require('thingsboard_api');
var myObjectsID = await TB.get.objectID('myObject', 'asset')
```
Result
```js
"ea791310-78d2-11ea-a1c7-d1e730c27b32"
```

### get.objectIDandKeys()
### Promise get ID and attributes of an object by its name,type and keys

List of options:
```json
(name, type, keys) //string. for keys can be array
//If keys == null - Trying to get all attributes!
```

Usage
```js
var TB = require('thingsboard_api');
var myObjectID = await TB.get.objectIDandKeys('myObject', 'asset', 'key1,key2,key3')

let keys = ['key1','key2','key3']
var myObjectIDandAttrs = await TB.get.objectIDandKeys('myObject', 'asset', keys)
```
Result
```js
{
    id: "ea791310-78d2-11ea-a1c7-d1e730c27b32", 
    name: "myObject", 
    type: "asset",
    key1: "value1",
    key2: "value2",
}
```




### get.allObjectsIDbyType()
### Promise get all object's ID by its name and "custom type",type

List of options:
```json
(type, entity_type) //string. type - custom type, entity_type - ASSET/DEVICE/ENTITY_VIEW
```

Usage
```js
var TB = require('thingsboard_api');
var allObjectsID = await TB.get.allObjectsIDbyType('device_type', 'device')

```
Result
```js
[
    {
        id: "0b507930-78d2-11ea-a1c7-d1e730c27b32", 
        name: "name1", 
        type: "device_type"
    },

    {
        id: "0b507930-78d2-11ea-a1c7-d1e730c27b32", 
        name: "name2", 
        type: "device_type"
    }
]
```

### get.allObjectsIDandKeysByType()
### Promise get all object's ID and attributes by its name and "custom type",type and keys

List of options:
```json
(type, entity_type, keys) //string. type - custom type, entity_type - ASSET/DEVICE/ENTITY_VIEW. For keys - can be array.
//If keys == null - Trying to get all attributes!
```

Usage
```js
var TB = require('thingsboard_api');
let keys = ['key1','key2','key3']
var allObjectsIDandAttrs = await TB.get.allObjectsIDandKeysByType('device_type', 'device',keys)

```

Result
```js
[
    {
        id: "0b507930-78d2-11ea-a1c7-d1e730c27b32", 
        name: "name1", 
        type: "device",
        key1: "value1",
        key2: "value2"
    },

    {
        id: "0b507930-78d2-11ea-a1c7-d1e730c27b32", 
        name: "name2", 
        type: "device",
        key1: "value1",
        key2: "value2"
    }
]
```


## PUSH

### push.createRelation()

List of options:
```json
(name, entity_view, parentName, parentType) //string
```

Usage
```js
var TB = require('thingsboard_api');
var createRelation = await TB.push.createRelation("myChild",'asset',"myParent",'asset')
```
Result
```js
true || false
```

### push.pushAttributes()

List of options:
```json
(name,entity_type,attributes) //string
```

Usage
```js
var TB = require('thingsboard_api');
let attributes = {
        key1: "value1",
        key2: "value2",
    }
var pushAttributes = await TB.push.pushAttributes("myObject",'asset',attributes)
```
Result
```js
true || false
```

### push.createEntity()

List of options:
```json
(name,type,attributes,entity_type,parentName,parentType,parentKeys,parentRelation)
// attributes - can be null, if no attributes to push
// parentName,parentType,parentKeys,parentRelation - can be null
// so no relation or cloning attributes will be done
//
// parentKeys - array of name of keys to clone from parent to new Entity
// parentRelation - boolean. If true - will call createRelation() to parentName
```

Usage
```js
var TB = require('thingsboard_api');
let attributes = {
        key1: "value1",
        key2: "value2",
    }
let parentKeys = ["keyTestParent","keyTestParent2"]

var createEntity = await TB.push.createEntity("myAsset",'testType',attributes,"asset","ParentDevice","device",parentKeys,true))
```
Result
```js
id || false //if no relation and parentClonning

{
    id: id,
    statusAttributes: true || false,
    statusRelation: true || false,
}
```


## DELETE

### delete.deleteEntity()

List of options:
```json
(name,entity_type,flagDeleteChilds) 
// flagDeleteChilds - boolean, if true - will delete child entities in 3 level depth max
```

Usage
```js
var TB = require('thingsboard_api');

var deleteEntity = await TB.delete.deleteEntity("myOldObject",'asset',false)
```
Result
```js
true || false
```

### delete.deleteEntitiesByType()

List of options:
```json
(type,entity_type) 

```

Usage
```js
var TB = require('thingsboard_api');

var deleteEntity = await TB.delete.deleteEntitiesByType("typeOfAssetsToDelete",'asset')
```
Result
```js
true || false
```


### delete.deleteEntitiesByType()

Will delete entities, whose has relation to 'name' in 3 level depth max
List of options:
```json
(name,entity_type) 

```

Usage
```js
var TB = require('thingsboard_api');

var deleteChilds = await TB.delete.deleteChilds("Parent",'asset')
```
Result
```js
true || false
```

# List Postgres Functions

### postgres.get.allObjectsIDbyType()
### Promise get all object's ID by its name and "custom type",type

List of options:
```json
(type, entity_type) //string. type - custom type, entity_type - ASSET/DEVICE/ENTITY_VIEW
```

Usage
```js
var TB = require('thingsboard_api');
var allObjectsID = await TB.postgres.get.allObjectsIDbyType('device_type', 'device')

```
Result
```js
[
    {
        id: "0b507930-78d2-11ea-a1c7-d1e730c27b32", 
        entity_name: "name1", 
        type: "device_type"
    },

    {
        id: "0b507930-78d2-11ea-a1c7-d1e730c27b32", 
        entity_name: "name2", 
        type: "device_type"
    }
]
```

### postgres.get.allObjectsIDandKeysByType()
### Promise get all object's ID and attributes by its name and "custom type",type and keys

List of options:
```json
(type, entity_type, keys) //string. type - custom type, entity_type - ASSET/DEVICE/ENTITY_VIEW. For keys - can be array.
//If keys == null - Trying to get all attributes!
```

Usage
```js
var TB = require('thingsboard_api');
let keys = ['key1','key2','key3']
var allObjectsIDandAttrs = await TB.postgres.get.allObjectsIDandKeysByType('device_type', 'device',keys)

```

Result
```js
[
    {
        entity_id: "0b507930-78d2-11ea-a1c7-d1e730c27b32", 
        entity_name: "name1", 
        entity_type: "device",
        key1: "value1",
        key2: "value2"
    },

    {
        entity_id: "0b507930-78d2-11ea-a1c7-d1e730c27b32", 
        entity_name: "name2", 
        entity_type: "device",
        key1: "value1",
        key2: "value2"
    }
]
```

___

#### postgres.get.getAttrsAndValuesById(entity_id) - get attributes keys and values for future assigning to child

##### Steps: 
 - convert Thingsboard UUID to Postgres entity_id
 - Use postgres.toPostgresID(thingsboard_uuid) for converting

##### List of options:
 - entity_id - is `string`

##### Response:
```js
[
...,
 {
    entity_type: 'DEVICE',
    entity_id: '1e9fbe382c16090a0332dde0dc34203',
    attribute_type: 'CLIENT_SCOPE',
    attribute_key: 'attribute_test_april2020',
    bool_v: false,
    str_v: '',
    long_v: 0,
    dbl_v: 0,
    last_update_ts: null
  },
...
]
```

____


#### postgres.toPostgresID(thingsboard_uuid) - convert Thingsboard UUID to Postgres ID

##### List of options:
 - thingsboard_uuid - `string`


___

#### postgres.insertIntoAttrsKeysVals(dataToWrite) - insert data into attribute_kv

##### List of options:
 - dataToWrite - is `array of objects`
```js
[
  {
    entity_type: 'DEVICE',
    entity_id: '1ea6d07aaba34b094de3ddf86487a77',
    attribute_type: 'CLIENT_SCOPE',
    attribute_key: 'test_attr_key',
    bool_v: null,
    str_v: null,
    long_v: null,
    dbl_v: null,
    last_update_ts: 1586949836446
  },
  {
    entity_type: 'DEVICE',
    entity_id: '1ea6d07aaba34b094de3ddf86487a77',
    attribute_type: 'CLIENT_SCOPE',
    attribute_key: 'Работа по программе',
    bool_v: null,
    str_v: null,
    long_v: null,
    dbl_v: null,
    last_update_ts: 1586949836446
  },
]
```

##### Response:
 - To check if data was written to db `compare insert count with response count`

___

### postgres.updateAttrsKeysAndVals(attributeObj) - Update child attributes keys and values based on parent attributes

##### List of options:
 - attributeObj is  `object`
```js
  {
    entity_type: 'DEVICE',
    entity_id: '1ea6d07aaba34b094de3ddf86487a77',
    attribute_type: 'CLIENT_SCOPE',
    attribute_key: 'Работа по программе',
    bool_v: null,
    str_v: null,
    long_v: null,
    dbl_v: null,
    last_update_ts: 1586949836446
  }
```
##### Response:
 - To check if data was updated `compare target count with db update response count`


# List of unclassified functions

### extendChildAttrs(options) - Extend child attributes by parent attributes

##### List of options:
 - options is an `object`
```js
{
  parent_id": "82c16090-fbe3-11e9-a033-2dde0dc34203",
  "child_id": "aaba34b0-6d07-11ea-94de-3ddf86487a77",
  "child_type": "DEVICE",
  "POSTGRES_HOST": "host",
  "POSTGRES_PORT": "port",
  "POSTGRES_USERNAME": "username",
  "POSTGRES_PASSWORD": "pass",
  "POSTGRES_DATABASE": "database",
  "updateAttrs": "false"
}
```

##### Steps:
 - If you want to add not existed attributes before to child set `updateAttrs` to `false`
 - For updating child attributes by parents attributes set `updateAttrs` to `true`

##### Reponse:
 - Function doesn't return any data!

CHANGELOG
Added Postgres functiones

[postgres.get.allObjectsIDbyType - Promise get all object's ID by its name and "custom type",type](https://www.npmjs.com/package/thingsboard_api#postgresgetallobjectsidbytype---promise-get-all-objects-id-by-its-name-and-custom-typetype)
[postgres.get.allObjectsIDandKeysByType - Promise get all object's ID and attributes by its name and "custom type",type and keys](https://www.npmjs.com/package/thingsboard_api#postgresgetallobjectsidandkeysbytype---promise-get-all-objects-id-and-attributes-by-its-name-and-custom-typetype-and-keys)

test.js
var TB = require("./index.js")
async function main(){
    let options = {
        TB_HOST:'84.201.141.244',
        TB_PORT:'8080',
        TB_USERNAME:'kp@2050.digital',
        TB_PASSWORD:'aPb79AsA' 
    }
    await TB.createConnection(options)
    var e5kObj = await TB.get.objectIDandKeys('e5k','asset',null)
    console.log(e5kObj)
    console.log(await TB.get.allObjectsIDandKeysByType('Тип локомотива', 'asset',null))
}

main()

test/test.js
var TB = require("../index.js")
async function main(){
    let options = {
        TB_HOST:'84.201.141.244',
        TB_PORT:'8080',
        TB_USERNAME:'kp@2050.digital',
        TB_PASSWORD:'aPb79AsA',
        POSTGRES_HOST:"84.201.141.244",
        POSTGRES_PORT:"5432",
        POSTGRES_USERNAME: "postgres",
        POSTGRES_PASSWORD:"postgres123456"
    }
    await TB.createConnection(options)
    var keys = ["errors", 'hardware', 'inspectionType', 'label', 'locoType','measureID', 'milage','name','ts','warnings','worker']


    var startSQL = new Date()
    var allMeasuresFull_SQL_nullKeys = await TB.postgres.get.allObjectsIDandKeysByType('Замер','entity_view',null)
    var timeSQL_nullKeys = (new Date()-startSQL)/1000;

    var startTB = new Date()
    var allMeasuresFull_TB_nullKeys = await TB.get.allObjectsIDandKeysByType('Замер','entity_view',null)
    var timeTB_nullKeys = (new Date()-startTB)/1000;


    var startSQL = new Date()
    var allMeasuresFull_SQL_allKeys = await TB.postgres.get.allObjectsIDandKeysByType('Замер','entity_view',keys)
    var timeSQL_allKeys = (new Date()-startSQL)/1000;

    var startTB = new Date()
    var allMeasuresFull_TB_allKeys = await TB.get.allObjectsIDandKeysByType('Замер','entity_view',keys)
    var timeTB_allKeys = (new Date()-startTB)/1000;

    var startSQL = new Date()
    var allMeasuresFull_SQL_oneKey = await TB.postgres.get.allObjectsIDandKeysByType('Замер','entity_view',"errors")
    var timeSQL_oneKey = (new Date()-startSQL)/1000;

    var startTB = new Date()
    var allMeasuresFull_TB_oneKey = await TB.get.allObjectsIDandKeysByType('Замер','entity_view',"errors")
    var timeTB_oneKey = (new Date()-startTB)/1000;


}

main()

test/testTB_get.js
/*
TB.get.objectID
TB.get.objectIDandKeys
TB.get.allObjectsIDbyType
TB.get.allObjectsIDandKeysByType
*/

var TB = require("../index.js")
const config = require('dotenv').config();

async function main() {
    // let options = {
    //     TB_HOST:'84.201.141.244',
    //     TB_PORT:'8080',
    //     TB_USERNAME:'kp@2050.digital',
    //     TB_PASSWORD:'aPb79AsA',
    //     POSTGRES_HOST:"84.201.141.244",
    //     POSTGRES_PORT:"5432",
    //     POSTGRES_USERNAME: "postgres",
    //     POSTGRES_PASSWORD:"postgres123456"
    // }
    // get params from env
    let options = {
        POSTGRES_HOST: config.parsed.PG_HOST,
        POSTGRES_PORT: config.parsed.PG_PORT,
        POSTGRES_USERNAME: config.parsed.PG_USERNAME,
        POSTGRES_PASSWORD: config.parsed.PG_PASSWORD,
        POSTGRES_DATABASE: config.parsed.PG_DATABASE,
        TB_HOST: config.parsed.TB_HOST,
        TB_PORT: config.parsed.TB_PORT,
        TB_USERNAME: config.parsed.TB_USERNAME,
        TB_PASSWORD: config.parsed.TB_PASSWORD,
        "entityId": "f6d6d160-7fd0-11ea-8b96-39ec393585a5",
        "entityType": "DEVICE",
        "entityName": "test-dev123321",
        "ts": Date.now(),
        "telemetry": {
            "someKey": "123",
            "anotherKey": "421",
        },
        "attributes": {
            "attr1": "val1",
            "attr2": "val2",
        }
    }

    await TB.createConnection(options)

    // var keys = ["errors", 'hardware', 'inspectionType', 'label', 'locoType','measureID', 'milage','name','ts','warnings','worker']
    /*
        var objectID = await TB.get.objectID("e5k",'asset')
        var objectIDandKeys = await TB.get.objectIDandKeys("Замер №1",'entity_view', null)
        var allObjectsIDbyType = await TB.get.allObjectsIDbyType("Локомотив","asset")
        var allObjectsIDandKeysByType = await TB.get.allObjectsIDandKeysByType("Замер","entity_view",keys)
        */
    //var related = await TB.get.relations("Замер №20&e5k:018-1", 'entity_view', 'to',1)

    // var related = await TB.get.relations("Замер №20", 'entity_view', 'to',3) 


    // last param set to get complex object info
    /*{
        id: "0b507930-78d2-11ea-a1c7-d1e730c27b32", 
        name: "some_name", 
        type: "device_type",
        deviceToken: "access_token",
    }*/

    // if it not set, response will be TB id
    try {
        const result = await TB.get.objectID(options.entityName, options.entityType, true);
        console.log(result);
    } catch (err) {
        console.error("Error: ", err);
    }

    // set last option to true
    // if your tb type is device
    // to get device token
    try {
        const result = await TB.get.allObjectsIDbyType("123", "device", true);
        console.log('get id by type resp: ', result);
    } catch(err){
        console.error("Error: ", err);
    }
}

main()

index.js
const fetch = require('node-fetch');
const axios = require('axios').default;
const cron = require('node-cron');
const postgres_api = require('./postgres.js');
const TB_get_api = require('./TB/get.js');
const funcs = require('./functions');
const TB_push_api = require('./TB/push.js');
// TB_HOST = 'localhost'
// TB_PORT = '8080'
// TB_USERNAME = 'tenant@thingsboard.org'
// TB_PASSWORD = 'tenant'
// process.env.POSTGRES_HOST  = 'localhost'
// process.env.POSTGRES_PORT = "5432"
// process.env.POSTGRES_USERNAME = 'postgres'
// process.env.POSTGRES_PASSWORD = 'postgres'

async function createConnection(options) {
    process.env.TB_HOST = options.TB_HOST;
    process.env.TB_PORT = options.TB_PORT
    process.env.TB_USERNAME = options.TB_USERNAME;
    process.env.TB_PASSWORD = options.TB_PASSWORD;
    process.env.POSTGRES_HOST = options.POSTGRES_HOST;
    process.env.POSTGRES_PORT = options.POSTGRES_PORT
    process.env.POSTGRES_USERNAME = options.POSTGRES_USERNAME;
    process.env.POSTGRES_PASSWORD = options.POSTGRES_PASSWORD;
    TB_HOST = process.env.TB_HOST;
    TB_PORT = process.env.TB_PORT
    TB_USERNAME = process.env.TB_USERNAME;
    TB_PASSWORD = process.env.TB_PASSWORD;
    await token()
    await postgres_api.createPostgresConnection();
}

async function token() {
    var url = 'http://' + process.env.TB_HOST + ':' + process.env.TB_PORT + '/api/auth/login';
    var options = {
        method: 'post',
        url: url,
        data: {
            "username": process.env.TB_USERNAME,
            "password": process.env.TB_PASSWORD
        },
        headers: {
            "Content-type": "application/json",
            "Accept": "application/json"
        }
    };

    var token = await getAndSetToken(options);
    cron.schedule("*/15 * * * *", async () => {
        await getAndSetToken(options);
    })
    return token

}

async function getAndSetToken(options) {
    try {
        const response = await axios(options);

        if (response.status === 200) {
            process.env.TB_TOKEN = response.data.token;
            return process.env.TB_TOKEN
        }
    } catch (error) {
        console.error(error);
    }
}

async function extendChildAttrs(options) {
    const parentId = postgres_api.toPostgresID(options.parent_id);
    const childId = postgres_api.toPostgresID(options.child_id);
    const childType = options.child_type;

    const parentAttrs = await postgres_api.get.getAttrsAndValuesById(parentId);
    const childAttrs = await postgres_api.get.getAttrsAndValuesById(childId);

    const parentAttrsValues = funcs.makeAttrsValuesObj(parentAttrs);
    const childAttrsValues = funcs.makeAttrsValuesObj(childAttrs);

    const parentKeys = Object.keys(parentAttrsValues);

    for (let parentKey of parentKeys) {
        // get attribute_types which are not find in child
        // the write new attribute_keys 
        if (!(childAttrsValues.hasOwnProperty(parentKey))) {
            console.log('child does not have attribute type: ', parentKey);
            let dataToWrite = parentAttrsValues[parentKey];
            // set child entity_id, entity_type for extending attributes of child
            // set null to attributes which not existed before extending of attributes
            dataToWrite = funcs.updateChildAttrsKeysValue(dataToWrite, childId, childType);

            const result = await postgres_api.insertIntoAttrsKeysVals(dataToWrite);

            if (result.count === dataToWrite.length) {
                console.log('successfully write to db!')
            }

            // child and parent have common attributes type SERVER_SCOPE and etc.
            // we need to find attribute_keys which not presented in child
        } else {
            console.log('parent and child have common attribute type ', parentKey);

            let dataToWrite = [];
            const parentData = parentAttrsValues[parentKey];
            const childData = childAttrsValues[parentKey];

            // update attributes or assign new attributes to child
            // depends on options.updateAttrs
            switch (options.updateAttrs) {
                case true:
                    if (parentData.length !== childData.length) {
                        console.error("Different length of parent and child attribute array!");
                        return;
                    }
                    let updatedAttrs = funcs.updateChildAttrsKeysValue(parentData, childId, childType);
                    for (let attr of updatedAttrs) {
                        const resp = await postgres_api.updateAttrsKeysAndVals(attr);
                        console.log('resp ', resp);
                    }
                    break;
                case false:
                default:
                    for (let i = 0; i < parentData.length; i++) {
                        let match = false;
                        for (let j = 0; j < childData.length; j++) {
                            if (parentData[i].attribute_key.toString() === childData[j].attribute_key.toString()) {
                                match = true;
                                break;
                            }
                        }
                        if (!match) {
                            // Change parent properties to child
                            parentData[i].entity_id = childId;
                            parentData[i].entity_type = childType;
                            parentData[i].last_update_ts = Date.now()
                            dataToWrite.push(parentData[i]);
                        }
                    }

                    if (dataToWrite.length === 0) {
                        console.log('Not data to write! ');
                        continue;
                    }

                    console.log('Find attributes to assign to child !');
                    console.log('data to write ', dataToWrite)

                    const result = await postgres_api.insertIntoAttrsKeysVals(dataToWrite);

                    if (result.count === dataToWrite.length) {
                        console.log('successfully write to db!')
                    }
            }
        }
    }
}
module.exports = {
    get: TB_get_api,
    push: TB_push_api,
    postgres: postgres_api,
    token: token,
    createConnection: createConnection,
    extendChildAttrs: extendChildAttrs,
};


TB/delete.js
const fetch = require('node-fetch');
const get = require('./get.js')


async function deleteEntity(name,entity_type,flagDeleteChilds){
    let id = await get.objectID(name,entity_type)
    
    var url= "http://" + TB_HOST + ':' + TB_PORT + "/api/"+entity_type.toLowerCase()+"/"+id
    if(entity_type.toLowerCase() == 'entity_view')
        url = "http://" + TB_HOST + ':' + TB_PORT + "/api/entityView/"+id


    let post = await fetch(url,{
        method: 'delete',
        headers: {
            'Content-Type': 'application/json',
            'X-Authorization': 'Bearer ' + process.env.TB_TOKEN
        }
    });

    var ans = await post.text();
    if(ans != '' )
        return false
    else if(ans == '' && (!flagDeleteChilds || flagDeleteChilds == null))
        return true

    var statusDeleteChilds = false
    if(flagDeleteChilds)
        statusDeleteChilds = await deleteChilds(name,entity_type)
    return true
}

async function deleteEntitiesByType(type,entity_type){
    entity_type = entity_type.toLowerCase()
    if(type == null || entity_type!='asset' || entity_type!='device' || entity_type!='entity_view')
        return false

    ids = get.allObjectsIDbyType(type,entity_type)
    for(let i = 0; i < ids.length; i++){
        var result = await deleteEntity(ids[i].name,ids[i].entity_type,false)
    }
    return result
}


async function deleteChilds(name,entity_type){

    var childs = await get.relations(name, entity_type, "to", 0)
    for (let i=0; i < childs.length; i++){
        await deleteEntity(childs[i].name, childs[i].entity_type,false)
    }
    return true
}

module.exports = {
    deleteEntity:           deleteEntity,
    deleteChilds:           deleteChilds,
    deleteEntitiesByType:   deleteEntitiesByType
};

functions.js
// Here usefull functions for work is presented

// obj which will store attribute keys according to attribute_type
/* {
    "SERVER_SCOPE": ["active, "простой",.....],
    "CLIENT_SCOPE": [...],
}
*/
function makeAttrsValuesObj(incomeAttrs){
    let outputObj = {};
    let buf = [];

    for (let i = 0; i < incomeAttrs.length; i++) {
        if (i === incomeAttrs.length - 1) {
            buf.push(incomeAttrs[i])
            outputObj[incomeAttrs[i].attribute_type] = buf;
            break;
        }

        if (incomeAttrs[i].attribute_type === incomeAttrs[i + 1].attribute_type) {
            buf.push(incomeAttrs[i]);
        }

        if (incomeAttrs[i].attribute_type !== incomeAttrs[i + 1].attribute_type) {
            buf.push(incomeAttrs[i]);
            outputObj[incomeAttrs[i].attribute_type] = buf;
            buf = [];
        }
    }

    return outputObj;
}

function updateChildAttrsKeysValue(childAttributes, childId, childType){
    // childAttributes is parent attributes that not existed before in child
    // before assign new attributes to child
    // change entity_id, entity_type and etc.
    for (let val of childAttributes) {
        val.entity_id = childId;
        val.entity_type = childType;
        val.bool_v = null;
        val.str_v = null;
        val.long_v = null;
        val.dbl_v = null;
        val.last_update_ts = Date.now();
    }

    return childAttributes;
}

module.exports = {
    makeAttrsValuesObj: makeAttrsValuesObj,
    updateChildAttrsKeysValue: updateChildAttrsKeysValue,
}

postgres.js
const postgres = require("postgres")

sqlConfig = {};

async function createPostgresConnection() {
    sqlConfig =
    {
        timeout: 1, // get idle connection
        evict: 2000, // it actualy removes the idle connection
        host: process.env.POSTGRES_HOST,         // Postgres ip address or domain name
        port: process.env.POSTGRES_PORT,       // Postgres server port
        database: process.env.PG_DATABASE,         // Name of database to connect to
        username: process.env.POSTGRES_USERNAME,         // Username of database user
        password: process.env.POSTGRES_PASSWORD,
    }

    return
}

function toUUID(id) {
    id = id.substring(7, 15) + "-" +
        id.substring(3, 7) + "-1" +
        id.substring(0, 3) + "-" +
        id.substring(15, 19) + "-" +
        id.substring(19);
    return id;
}

// get postgres id from thingsboard uuid
function toPostgresID(tb_uuid) {
    id = tb_uuid.substring(15, 18) +
        tb_uuid.substring(9, 13) +
        tb_uuid.substring(0, 8) +
        tb_uuid.substring(19, 23) +
        tb_uuid.substring(24, tb_uuid.length)

    return id;
}

// Get access token for device
async function getEntityToken(entityId){
    const sql = postgres('postgres://username:password@host:port/database', sqlConfig);
    try{
        const pgId = toPostgresID(entityId);
        var entityToken = await sql`SELECT credentials_id FROM device_credentials WHERE device_id = ${pgId}`
    } catch(error){
        console.error(error);
    }
    
    return entityToken[0].credentials_id;
}
async function getAttrsAndValuesById(entityId) {
    const sql = postgres('postgres://username:password@host:port/database', sqlConfig);
    try {
        var outputAttrs = await sql`SELECT entity_type, entity_id, attribute_type, attribute_key, bool_v, str_v, long_v, dbl_v, last_update_ts FROM attribute_kv where entity_id = ${entityId} ORDER BY attribute_type DESC`;
    } catch (error) {
        console.error(error);
    }

    return outputAttrs;
}

async function insertIntoAttrsKeysVals(dataToWrite) {
    console.log('data ', dataToWrite)
    const sql = postgres('postgres://username:password@host:port/database', sqlConfig);
    try {
        var insertResponse = await sql`insert into attribute_kv ${sql(dataToWrite,
            'entity_type', 'entity_id', 'attribute_type', 'attribute_key', 'bool_v', 'str_v', 'long_v', 'dbl_v', 'last_update_ts'
        )}`;
    } catch (error) {
        console.error(error);
    }

    return insertResponse;
}

// Update child attributes keys and values based on parent attributes
async function updateAttrsKeysAndVals(attributeObj) {
    const sql = postgres('postgres://username:password@host:port/database', sqlConfig);
    try {
        var updateResponse = await sql`update attribute_kv set ${
            sql(attributeObj, 'entity_type', 'entity_id', 'attribute_type', 'attribute_key', 'bool_v', 'str_v', 'long_v', 'dbl_v', 'last_update_ts')    
        } where entity_id = ${ attributeObj["entity_id"] } and attribute_key = ${attributeObj.attribute_key}`;
    } catch (error) {
        console.error(error);
    }

    return updateResponse;
}

async function getAllObjectsIDbyType(type, entity_type) {
    const sql = postgres('postgres://username:password@host:port/database', sqlConfig)

    const data = await sql`
        SELECT name,entity_type,id from ${sql(entity_type)}
        WHERE type = ${type} ORDER BY id DESC `

    for (let i = 0; i < data.length; i++) {
        data[i].id = toUUID(data[i].id)
    }

    return data
}

async function getAllObjectsIDandKeysByType(type, entity_type, keys) {
    const sql = postgres('postgres://username:password@host:port/database', sqlConfig)

    const data = await sql`
        SELECT name,entity_type,id from ${sql(entity_type)}
        WHERE type = ${type} ORDER BY id DESC `

    var entities = []
    for (let i = 0; i < data.length; i++) {
        entities.push(data[i].id)
    }
    var result
    if (keys != null) {
        result = await sql`
            SELECT entity_id, entity_type, long_v, str_v, attribute_key from public.attribute_kv
            WHERE  entity_id in (${entities}) and 
                attribute_key in (${keys})
            ORDER by entity_id DESC
            `;
    }
    else {
        result = await sql`
        SELECT entity_id, entity_type, long_v, str_v, attribute_key from public.attribute_kv
        WHERE  entity_id in (${entities})
        ORDER by entity_id DESC
        `;
    }
    let target = []
    let obj = {}
    if (typeof timezoneOffset == 'undefined')
        var timezoneOffset = 0
    for (let i = 0; i < result.length; i++) {
        if (result[i]["attribute_key"] == 'ts')
            result[i]["long_v"] = new Date(result[i]["long_v"] - timezoneOffset * 60 * 1000)


        if (i === result.length - 1) {
            obj[result[i]["attribute_key"]] = result[i]["long_v"] || result[i]["str_v"];
            obj["entity_id"] = result[i]["entity_id"].toString();
            obj["entity_type"] = result[i]["entity_type"].toString();
            target.push(obj);
            break;
        }

        if (result[i]["entity_id"].toString() === result[i + 1]["entity_id"].toString()) {


            obj[result[i]["attribute_key"]] = result[i]["long_v"] || result[i]["str_v"];
            if (result[i]["long_v"] === 0)
                obj[result[i]["attribute_key"]] = 0;
        }

        if (result[i]["entity_id"].toString() !== result[i + 1]["entity_id"].toString()) {
            obj[result[i]["attribute_key"]] = result[i]["long_v"] || result[i]["str_v"];
            obj["entity_id"] = toUUID(result[i]["entity_id"].toString());
            obj["entity_type"] = result[i]["entity_type"].toString();
            obj["entity_name"] = data[target.length].name
            target.push(obj)
            obj = {};
        }
    }

    return target
}


module.exports =
{
    createPostgresConnection: createPostgresConnection,
    toPostgresID: toPostgresID,
    insertIntoAttrsKeysVals: insertIntoAttrsKeysVals,
    updateAttrsKeysAndVals: updateAttrsKeysAndVals,
    get: {
        allObjectsIDbyType: getAllObjectsIDbyType,
        allObjectsIDandKeysByType: getAllObjectsIDandKeysByType,
        getAttrsAndValuesById: getAttrsAndValuesById,
        getEntityToken: getEntityToken,
    }
}